src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -44,12 +50,18 @@ public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, str
             => CreateAndAddToWorkspace(projectSystemName, language, new VisualStudioProjectCreationInfo());
 
         public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo)
+            => _threadingContext.JoinableTaskFactory.Run(async () => await CreateAndAddToWorkspaceAsync(
+                projectSystemName, language, creationInfo, CancellationToken.None).ConfigureAwait(false));
+
+        public async Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo, CancellationToken cancellationToken)
         {
             // HACK: Fetch this service to ensure it's still created on the UI thread; once this is moved off we'll need to fix up it's constructor to be free-threaded.
             _visualStudioWorkspaceImpl.Services.GetRequiredService<VisualStudioMetadataReferenceManager>();
 
-            // HACK: since we're on the UI thread, ensure we initialize our options provider which depends on a UI-affinitized experimentation service
-            _visualStudioWorkspaceImpl.EnsureDocumentOptionProvidersInitialized();
+            await _visualStudioWorkspaceImpl.EnsureDocumentOptionProvidersInitializedAsync(cancellationToken).ConfigureAwait(false);
+
+            // From this point on, we start mutating the solution.  So make us non cancellable.
+            cancellationToken = default;
```

Comment:
I like this pattern. You may wish to consider fetching the IVsSolution earlier as well so that way any of the "async" stuff is done up front.

Reply:
i tried this... but it didn't seem to improve things:

1. we only want the solution in the case that ProjectIds.Count == 0.  
2. we still call IVsSolution.GetSolutionInfo, which i imagine may need the UI thread.  So i think we still need to switch over to be on the safe side.

If you have an alternative for this, i'm all ears.

Reply:
acutally, did what you asked since we are always on hte UI thread.  it prsupposes that getting the solutoin object and solution filepath is cheap.  but that seems somewhat reasonable to assume.